### DISCUSSION TITLE
Import Styling

### DISCUSSION THREAD

**Discussion by asahasrabuddhe** [go-fuego/fuego#184] [General]
In some of the Go projects, I have worked on, we have grouped the imports by their source. For example,

```Go
import (
    // std lib imports

    // third party imports

    // package imports from our own repo
)
```

So, the import section for `fuego/cmd/fuego/commands/controller.go` which currently looks like:

```Go
import (
	"fmt"
	"os"
	"strings"

	"github.com/go-fuego/fuego/cmd/fuego/templates"
	"github.com/urfave/cli/v2"
	"golang.org/x/text/cases"
	"golang.org/x/text/language"
)
```

would then be written as:

```Go
import (
	"fmt"
	"os"
	"strings"

	"github.com/urfave/cli/v2"
	"golang.org/x/text/cases"
	"golang.org/x/text/language"

	"github.com/go-fuego/fuego/cmd/fuego/templates"
)
```

What do you think?

**Comment by EwenQuim** [comment]
We currently just use `gofmt` for all modules and I sometimes force gofumpt on the main package (`gofumpt` is steicter than `go fmt`).

One of the enjoyable things in Go is to have only one formatter. You can change to import order but it must still be compatible with the formatters.

**Reply by ccoVeille** [reply]
I'm unsure about the argument here.

gofmt format the code, yes

But goimport is the one that formats the imports.

gofmt doesn't reorder the import for example, by order I mean `fmt` package import before `strings`. Just try it in command gofmt on file with messed up imports, or a file were you add a dependency to a new package. It will do nothing.

Often goimport is invisible because embedded in gopls via the IDE

Also goimport is very basic. By default, there is nothing about grouping things. So goimport will only fix the missing imports, remove the unused ones, and reorder imports in each block. Which means that if you manually split in std lib, external+module, it will respect it. But if you do std,external, module it would also. The weirdest is that if you group std+external+module, goimport will sort them in the same group without separating them.

goimport has a -local argument to provide what it suggested in the question.

This argument can be passed to gopls via formating.local

https://github.com/golang/tools/blob/master/gopls%2Fdoc%2Fsettings.md#local-string

The initial question is legitimate. Many people and projects use the "third party" section to separate the import related to the local module. I also support this way to do.

That said, I can heard the argument of "that how Go standard tools handle all these without any configuration", that sounds like a "thanks, but no." reply to the question. I'm fine with it.

My 2 cents,

Respectfully, Christophe

**Reply by EwenQuim** [reply]
Oh ok, I just use my IDE.

If you want to format the imports, I'm ok with that, you can send a PR, it will be accepted :)

**Reply by ccoVeille** [reply]
I provided a fix with
- https://github.com/go-fuego/fuego/pull/186

**Comment by ccoVeille** [comment]
I consider this as the norm now.

goimport can be easily configured with the -local argument

gopls also supports it via formating.local settings.

I tend to use [`gci`](https://github.com/daixiang0/gci) to format the import, as it provides a better control on the formatting, even more with the -localmodule argument that is an easy way to format the local module without having to pass an argument. So you can use the same setting for all projects.

So, yes I support your suggestion.

But OSS is a matter of discussion.

Thanks for raising the point, I was also surprised to see fuego using a "mixed up import".

**Comment by asahasrabuddhe** [comment]
Closing the discussion as #186 has addressed my concern.
